protocol/contracts/v2/circuitBreaker/TempleCircuitBreakerAllUsersPerPeriod.sol

```diff
@@ -0,0 +1,169 @@
+        uint256 _cap
+    ) TempleElevatedAccess(_initialRescuer, _initialExecutor) {
+        _setConfig(_periodDuration, _nBuckets, _cap);
+    }
+
+    /**
+     * @notice Verify the new amount requested does not breach the cap in this rolling period.
+     */
+    function preCheck(address /*onBehalfOf*/, uint256 amount) external override onlyElevatedAccess {
+        uint32 _nextBucketIndex = uint32(block.timestamp / secondsPerBucket);
+        uint32 _bucketIndex = bucketIndex;
+        uint32 _nBuckets = nBuckets;
+        
+        // If this time bucket is different to the last one
+        // then delete any buckets in between first - that is old data
+        if (_nextBucketIndex != _bucketIndex) {
+            unchecked {
+                // If it was more than a full periodDuration ago, then all buckets are deleted.
+                uint32 _oneperiodDurationAgoIndex = _nextBucketIndex - _nBuckets;
+                uint32 i = _bucketIndex < _oneperiodDurationAgoIndex ? _oneperiodDurationAgoIndex : _bucketIndex;
+                for (; i < _nextBucketIndex; ++i) {
+                    // Set to dust
+                    buckets[(i+1) % _nBuckets] = 1;
+                }
+            }
+
+            bucketIndex = _nextBucketIndex;
+        }
+
+        unchecked {
+            buckets[_nextBucketIndex % _nBuckets] += amount;
+        }
+        
+        uint256 _newUtilisation = _currentUtilisation(_nBuckets);
+        if (_newUtilisation > cap) revert CapBreached(_newUtilisation, cap);
+    }
+
+    /**
+     * @notice Set the duration, buckets and cap. This will reset the clock for any totals
+     * added since in the new periodDuration.
+     */
+    function setConfig(uint256 _periodDuration, uint256 _nBuckets, uint256 _cap) external onlyElevatedAccess {
+        _setConfig(_periodDuration, _nBuckets, _cap);
+    }
+
+    /**
+     * @notice Update the cap for this circuit breaker
+     */
+    function updateCap(uint256 newCap) external onlyElevatedAccess {
+        cap = newCap.encodeUInt128();
+        emit CapSet(newCap);
+    }
+    
+    /**
+     * @notice What is the total utilisation so far in this `periodDuration`
+     */
+    function currentUtilisation() external view returns (uint256 amount) {
+        return _currentUtilisation(nBuckets);
+    }
+
+    function _currentUtilisation(uint256 _nBuckets) internal view returns (uint256 amount) {
+        // Unchecked is safe here because we know previous entries are under the cap.
+        unchecked {
+            for (uint256 i = 0; i < _nBuckets; ++i) {
+                amount += buckets[i];
+            }
+
+            // Remove the dust
+            amount -= _nBuckets;
+        }
+    }
+
+    function _setConfig(uint256 _periodDuration, uint256 _nBuckets, uint256 _cap) internal {
+        if (_periodDuration == 0) revert CommonEventsAndErrors.ExpectedNonZero();
+        if (_periodDuration % _nBuckets != 0) revert CommonEventsAndErrors.InvalidParam();
```

Comment:
could there be a global constant `minimumPeriodDuration`? you know, to protect against misconfiguration of seconds

Reply:
hmm i don't think it's warranted. The worst that can happen is that trades are blocked and it can be reconfigured again to fix. not a big deal